rainier-core/src/main/scala/com/stripe/rainier/sampler/Ehmc.scala

```diff
@@ -0,0 +1,50 @@
+package com.stripe.rainier.sampler
+
+import scala.collection.mutable.ListBuffer
+import com.stripe.rainier.compute._
+import com.stripe.rainier.core._
+
+/**
+  * Empirical HMC - automated tuning of step size and number of leapfrog steps
+  * @param l0 the initial number of leapfrog steps to use during the dual averaging phase
+  * @param k the number of iterations to use when determining the
+  * empirical distribution of the total number of leapfrog steps until a u-turn
+  */
+final case class EHMC(l0: Int, k: Int = 2000) extends Sampler {
+  def sample(density: DensityFunction,
+             warmupIterations: Int,
+             iterations: Int,
+             keepEvery: Int)(implicit rng: RNG): List[Array[Double]] = {
+
+    val lf = LeapFrog(density)
+    val params = lf.initialize
+    val stepSize =
+      DualAvg.findStepSize(lf, params, 0.65, l0, warmupIterations)
+
+    val empiricalL: Vector[Int] = lf.empiricalLongestSteps(l0, k, stepSize)
+
+    def discreteUniform(min: Int, max: Int)(implicit rng: RNG) = {
+      math.floor(rng.standardUniform * (max - min) + min).toInt
+    }
+
+    if (stepSize == 0.0)
+      List(lf.variables(params))
+    else {
+      val buf = new ListBuffer[Array[Double]]
+      var i = 0
+
+      while (i < iterations) {
+
+        // val j: Int = Categorical.list(Seq.fill(k)(1.0 / k)).generator.get
```

Comment:
I've changed my mind about using `Categorical` here since that would introduce a dependency between `sampler` and `core` which I've been trying to avoid.

However I've just noticed that `RNG` already has an `int` method which you can use here: `val j = rng.int(k)`.